File: src/main/java/ukalus/persistence/file/FileLog.java
/*
 * Created on May 13, 2004
 *  
 */
package ukalus.persistence.file;

import ukalus.persistence.Log;
import ukalus.persistence.Reference;
import ukalus.persistence.SerializedObjectIterator;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.ObjectOutputStream;
import java.io.RandomAccessFile;
import java.util.Iterator;
import java.util.function.Function;

/**
 * @author gremopm
 * 
 */
public class FileLog implements Log {

  private RandomAccessFile channel;

  public FileLog(File file) throws FileNotFoundException {
    this.channel = new RandomAccessFile(file, "rw");
  }

  public void clear() throws IOException {
    channel.setLength(0);
  }

  public void add(Function o) throws IOException {
    ByteArrayOutputStream buffer = new ByteArrayOutputStream();
    ObjectOutputStream stream = new ObjectOutputStream(buffer);
    stream.writeObject(o);
    byte[] data = buffer.toByteArray();
    channel.seek(channel.length());
    channel.writeInt(data.length);
    channel.write(data);
  }

  public Iterator<Function> iterator() {
    Iterator<Function> result;
    try {
      channel.seek(0);
      result = new SerializedObjectIterator<>(channel);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return result;
  }

  public void close() throws IOException {
    channel.close();
  }
}

File: src/main/java/ukalus/persistence/file/FileLog.kt
/*
 * Created on May 13, 2004
 *  
 */
package ukalus.persistence.file

import ukalus.persistence.Log
import ukalus.persistence.SerializedObjectIterator

import java.io.*

/**
 * @author gremopm
 */
class FileLog<T>(file: File) : Log<T> {

    private val channel = RandomAccessFile(file, "rw")

    override fun clear() {
        channel.setLength(0)
    }

    override fun add(o: T) {
        val buffer = ByteArrayOutputStream()
        val stream = ObjectOutputStream(buffer)
        stream.writeObject(o)
        val data = buffer.toByteArray()
        channel.seek(channel.length())
        channel.writeInt(data.size)
        channel.write(data)
    }

    override fun iterator(): Iterator<T> {
        try {
            channel.seek(0)
            return SerializedObjectIterator(channel)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    @Throws(IOException::class)
    override fun close() {
        channel.close()
    }
}